src/bundle.c

```diff
@@ -1625,6 +1625,136 @@ gboolean check_bundle_payload(RaucBundle *bundle, GError **error)
 	return res;
 }
 
+gboolean replace_signature(RaucBundle *bundle, const gchar *insig, const gchar *outpath, gboolean verify, GError **error)
+{
+	g_autoptr(GFile) outfile = NULL;
+	g_autoptr(RaucBundle) ibundle = NULL;
+	g_autoptr(GFile) ifile = NULL;
+	GFileOutputStream* istream = NULL;
+	g_autofree gchar *iname = NULL;
+	g_autofree gchar *tmpdir = NULL;
+	g_autoptr(GBytes) sig = NULL;
+	GError *ierror = NULL;
+	gboolean res = FALSE;
+	gssize size;
+
+	g_return_val_if_fail(bundle != NULL, FALSE);
+	g_return_val_if_fail(outpath != NULL, FALSE);
+	g_return_val_if_fail(insig != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	r_context_begin_step("replace_signature", "Replacing bundle signature", 1);
+
+	tmpdir = g_dir_make_tmp("rauc-replace-XXXXXX", &ierror);
+	if (tmpdir == NULL) {
+		g_propagate_prefixed_error(error, ierror,
+				"Failed to create tmp dir: ");
+		res = FALSE;
+		goto out;
+	}
+
+	iname = g_build_filename(tmpdir, "_bundle.raucb", NULL);
+	ifile = g_file_new_for_path(iname);
+	istream = g_file_create(ifile, G_FILE_CREATE_PRIVATE, NULL, &ierror);
+	if (istream == NULL) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to create the temporary bundle to replace signature: ");
+		goto out;
+	}
+
+	sig = read_file(insig, &ierror);
+	if (!sig) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to read signature file: ");
+		res = FALSE;
+		goto out;
+	}
+
+	res = g_seekable_seek(G_SEEKABLE(bundle->stream),
+			0, G_SEEK_SET, NULL, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to seek to start of original bundle: ");
+		goto out;
+	}
+
+	size = g_output_stream_splice(
+			(GOutputStream*)istream,
+			bundle->stream,
+			G_OUTPUT_STREAM_SPLICE_NONE,
+			NULL, &ierror);
+	if (size == -1) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"failed to append original bundle payload to temporary bundle: ");
+		res = FALSE;
+		goto out;
+	}
+
+	res = g_seekable_seek(G_SEEKABLE((GOutputStream*)istream),
```

Comment:
Doesn't this need a `g_seekable_truncate()` before seeking to the end? Otherwise you could have leftover parts from the old signature if it was larger.